substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1982,85 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Reset the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to reset a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * The reset ledger must become either a nominator or validator.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::reset_ledger())]
+		pub fn reset_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
+			maybe_total: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			let controller = Bonded::<T>::get(&stash).ok_or(Error::<T>::NotStash)?;
```

Comment:
if this controller is same as stash, we can return early as well (I think)?

Reply:
I think we could, but that's basically the same as the check in the line below. The check below is more comprehensive though.